### PR TITLE
Add code insights usage statistics to pings

### PR DESCRIPTION
**Pull request by tjkandala**
Progress

- [x] Track view events
- [x] Track UI interaction events (hover, resize)
- [x] Track insight creations, deletions, and edits 
- [x] Implement insight viewing queries
- [x] Implement insight creation queries
- [x] Test pings queries
- [x] Add to documentation, site admin page, and CHANGELOG
- [x] Open BigQuery schema PR (https://github.com/sourcegraph/analytics/pull/97)

### Web app changes

Observe settings updates and diff insights changes. Add user interaction event listeners to extension-contributed views. Both of these changes will likely be rewritten in the near future when the code insights creation UI is built.

### Metrics collection not implemented 

- Total number of lines of code and total number of repos encompassed in each insight
- Average time for insights to load (GraphQL request performance)
- Cache hit %

These metrics are difficult to collect before the code insights backend is in place

### Query plans

<details>
<summary>Viewing metrics query</summary>
<pre> Aggregate  (cost=590.15..590.16 rows=1 width=48) (actual time=0.296..0.297 rows=1 loops=1)
   -&gt;  Bitmap Heap Scan on event_logs  (cost=19.08..582.75 rows=185 width=69) (actual time=0.068..0.205 rows=182 loops=1)
         Recheck Cond: (name = ANY (&apos;{ViewInsights,InsightHover,InsightUICustomization,InsightDataPointClick}&apos;::text[]))
         Heap Blocks: exact=154
         -&gt;  Bitmap Index Scan on event_logs_name  (cost=0.00..19.03 rows=185 width=0) (actual time=0.047..0.047 rows=182 loops=1)
               Index Cond: (name = ANY (&apos;{ViewInsights,InsightHover,InsightUICustomization,InsightDataPointClick}&apos;::text[]))
 Planning Time: 0.146 ms
 Execution Time: 0.353 ms
(8 rows)
</pre>
</details>

<details>
<summary>Creation metrics query</summary>
<pre> Aggregate  (cost=76.23..76.24 rows=1 width=72) (actual time=0.100..0.101 rows=1 loops=1)
   -&gt;  Bitmap Heap Scan on event_logs  (cost=22.18..75.07 rows=14 width=97) (actual time=0.055..0.070 rows=13 loops=1)
         Recheck Cond: (name = ANY (&apos;{InsightEdit,InsightAddition,InsightRemoval,InsightConfigureClick,InsightAddMoreClick}&apos;::text[]))
         Heap Blocks: exact=13
         -&gt;  Bitmap Index Scan on event_logs_name  (cost=0.00..22.18 rows=14 width=0) (actual time=0.048..0.049 rows=13 loops=1)
               Index Cond: (name = ANY (&apos;{InsightEdit,InsightAddition,InsightRemoval,InsightConfigureClick,InsightAddMoreClick}&apos;::text[]))
 Planning Time: 0.158 ms
 Execution Time: 0.159 ms
(8 rows)
</pre>
</details>

<details>
<summary>Weekly first time creators query</summary>
<pre> Aggregate  (cost=20.12..20.13 rows=1 width=16) (actual time=0.053..0.054 rows=1 loops=1)
   -&gt;  GroupAggregate  (cost=20.02..20.11 rows=1 width=32) (actual time=0.049..0.050 rows=0 loops=1)
         Group Key: event_logs.anonymous_user_id
         Filter: (min(event_logs.&quot;timestamp&quot;) &gt; date_trunc(&apos;week&apos;::text, now()))
         Rows Removed by Filter: 1
         -&gt;  Sort  (cost=20.02..20.03 rows=4 width=32) (actual time=0.042..0.043 rows=4 loops=1)
               Sort Key: event_logs.anonymous_user_id
               Sort Method: quicksort  Memory: 25kB
               -&gt;  Bitmap Heap Scan on event_logs  (cost=4.44..19.98 rows=4 width=32) (actual time=0.026..0.032 rows=4 loops=1)
                     Recheck Cond: (name = &apos;InsightAddition&apos;::text)
                     Heap Blocks: exact=4
                     -&gt;  Bitmap Index Scan on event_logs_name  (cost=0.00..4.44 rows=4 width=0) (actual time=0.020..0.020 rows=4 loops=1)
                           Index Cond: (name = &apos;InsightAddition&apos;::text)
 Planning Time: 0.181 ms
 Execution Time: 0.089 ms
(15 rows)
</pre>
</details>

### CODE REVIEW

File: internal/types/types.go [hunk 1]
```diff
@@ -1321,7 +1321,7 @@ type SearchOnboarding struct {
 // Weekly usage statistics for the extensions platform
 type ExtensionsUsageStatistics struct {
 	WeekStart                  time.Time
-	UsageStatisticsByExtension map[string]*ExtensionUsageStatistics
+	UsageStatisticsByExtension *[]*ExtensionUsageStatistics
```
Comment (ryanslade) [inline internal/types/types.go:1324]: Why make this a pointer to s slice?

**Review by ryanslade — APPROVED** [review]
Backend LGTM

File: internal/types/types.go (hunk 1 above)
Comment (tjkandala) [reply internal/types/types.go:1324]: I wasn't able to find a way to use dynamic keys (in this case, extension IDs) in BigQuery records. @ebrodymoore is [this](https://github.com/sourcegraph/analytics/pull/97/commits/85d20a3b6f08b880e48f1bec15d1f9c780a3005e#diff-33710a3aa606c9a1871982c6f91692b8c9621ee6f41be408ec0649f4f42ca469R4077) an acceptable schema, or is there a better way to do this?

Comment (ryanslade) [reply internal/types/types.go:1324]: I'm not sure I follow. Slices are already pointers so I'm not sure how the extra level of indirection helps?

File: client/web/src/views/ViewContent.tsx [hunk 2]
```diff
@@ -20,42 +22,94 @@ export interface ViewContentProps
...
+        VersionContextProps,
+        TelemetryProps {
     viewContent: View['content']
+    viewID: string
     location: H.Location
     history: H.History
     globbing: boolean
+
+    /** To get container to track hovers for pings */
+    containerClassName?: string
```
Comment (felixfbecker) [inline client/web/src/views/ViewContent.tsx:34]: Passing a CSS class for this seems a bit hacky. Can't we pass an element reference?

File: client/web/src/insights/analytics.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
+import { isEqual } from 'lodash'
+import { isSettingsValid, Settings, SettingsCascadeOrError } from '../../../shared/src/settings/settings'
+import { TelemetryService } from '../../../shared/src/telemetry/telemetryService'
+
+const insightTypes = ['searchInsights', 'codeStatsInsights'] as const
```
Comment (felixfbecker) [inline client/web/src/insights/analytics.ts:5]: Why only these two? Here are all the extensions currently contributing insights: https://sourcegraph.com/extensions?query=category%3AInsights (plus I think your Snyk extension)

File: client/web/src/insights/analytics.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+function parseInsightSettingsKey(key: string): Omit<InsightData, 'configuration'> | undefined {
+    const insightType = getInsightTypeFromSettingsKey(key)
+
+    if (!insightType) {
+        return undefined
+    }
+
+    switch (insightType) {
+        case 'searchInsights': {
+            // example key: "searchInsights.insights.graphQLTypesMigration"
```
Comment (felixfbecker) [inline client/web/src/insights/analytics.ts:96]:
```suggestion
            // example key: "searchInsights.insight.graphQLTypesMigration"
```

File: client/web/src/SourcegraphWebApp.tsx [hunk 5]
```diff
@@ -300,6 +301,15 @@ class ColdSourcegraphWebApp extends React.Component<SourcegraphWebAppProps, Sour
...
         )
 
+        // Observe settings mutations for analytics
+        this.subscriptions.add(
+            from(this.platformContext.settings)
+                .pipe(bufferCount(2, 1))
+                .subscribe(([oldSettings, newSettings]) => {
+                    logCodeInsightsChanges(oldSettings, newSettings, eventLogger)
+                })
+        )
```
Comment (felixfbecker) [inline client/web/src/SourcegraphWebApp.tsx:311]: **question**: Are we also logging the total amount of insights in the backend, or only the additions/removals from the client?

**thought**: One disadvantage of doing this in the frontend, as opposed to in the backend intercepting settings mutations, is that we won't capture changes to configuration done through the API.

**Review by felixfbecker — APPROVED** [review]
Nice work!

File: internal/types/types.go (hunk 1 above)
Comment (tjkandala) [reply internal/types/types.go:1324]: My mistake, I thought you were referring to the change from a map to a slice. Thanks, I'll change this

File: CHANGELOG.md [hunk 6]
```diff
@@ -19,6 +19,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Structural search now supports searching indexed branches other than default. [#17726](https://github.com/sourcegraph/sourcegraph/pull/17726)
 - Pings now contain Sourcegraph extension activation statistics. [#16421](https://github.com/sourcegraph/sourcegraph/pull/16421)
 - Pings now contain aggregate Sourcegraph extension activation statistics: the number of users and number of activations per (public) extension per week, and the number of total extension users per week and average extensions activated per user. [#16421](https://github.com/sourcegraph/sourcegraph/pull/16421)
+- Pings now contain aggregate code insights usage data: totals insight views, interactions, edits, creations, and removals, and the count of unique users that view and create insights. [#16421](https://github.com/sourcegraph/sourcegraph/pull/17805)
```
Comment (Joelkw) [inline CHANGELOG.md:22]:
```suggestion
- Pings now contain aggregate code insights usage data: total insight views, interactions, edits, creations, removals, and counts of unique users that view and create insights. [#16421](https://github.com/sourcegraph/sourcegraph/pull/17805)
```

**Review by Joelkw — APPROVED** [review]
So excited to start tracking this  – thank you for implementing these!

**Review by unknwon — APPROVED** [review]
LGTM other than @ryanslade's comment about "pointer to a slice", and thanks for putting up the query plans in advance ❤️

**Comment by ebrodymoore** [comment]
@tjkandala could you respond to [this comment](https://github.com/sourcegraph/analytics/pull/97#issuecomment-773592201)? If it's possible it'd be great to change for consistency with other pings, but no worries if it's a non-trivial amount of work. Then I'll be able to get the schema into prod and we can merge this PR!

**Comment by tjkandala** [comment]
@ebrodymoore Yeah, I'm about to make changes to the code insights schema as well (array of insights by type like with extensions). I'll get those both done at once and ping you (might need to quickly review that the new changes are OK)

File: client/web/src/views/ViewContent.tsx (hunk 2 above)
Comment (tjkandala) [reply client/web/src/views/ViewContent.tsx:33]: We'd need to extract the `<div>`s [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/repo/tree/ViewGrid.tsx#L56-69) into a component to get individual refs, but react-grid-layout expects that all children are HTML elements. There are [ways to work around this](https://github.com/STRML/react-grid-layout/issues/299#issuecomment-524959229), but the positioning was still a little off when I tried it as I may have missed forwarding some props that weren't discussed in that issue.

I would consider relying on [undocumented](https://github.com/STRML/react-grid-layout/issues/1166) modified props from [`React.cloneElement`](https://sourcegraph.com/github.com/STRML/react-grid-layout/-/blob/lib/GridItem.jsx#L619-620) (how do we type the props?) to be hackier than relying on DOM APIs. I do agree that ideally we'd be passing a ref

**Comment by tjkandala** [comment]
"Viewing metrics query" and "creation metrics query" are now "platform query" (metrics regarding code insights in general) and "metrics by insight query" in order to support arbitrary insight types.

Here are the new query plans:

<details>
<summary>Platform query</summary>
<pre> Aggregate  (cost=378.19..378.20 rows=1 width=40) (actual time=0.270..0.271 rows=1 loops=1)
   -&gt;  Bitmap Heap Scan on event_logs  (cost=18.50..374.68 rows=108 width=50) (actual time=0.066..0.196 rows=116 loops=1)
         Recheck Cond: (name = ANY (&apos;{ViewInsights,InsightAddition,InsightConfigureClick,InsightAddMoreClick}&apos;::text[]))
         Heap Blocks: exact=112
         -&gt;  Bitmap Index Scan on event_logs_name  (cost=0.00..18.47 rows=108 width=0) (actual time=0.047..0.048 rows=116 loops=1)
               Index Cond: (name = ANY (&apos;{ViewInsights,InsightAddition,InsightConfigureClick,InsightAddMoreClick}&apos;::text[]))
 Planning Time: 0.145 ms
 Execution Time: 0.320 ms
(8 rows)
</pre>
</details>

<details>
<summary>Metrics by insight query</summary>
<pre> HashAggregate  (cost=390.62..391.96 rows=107 width=80) (actual time=0.264..0.266 rows=4 loops=1)
   Group Key: (argument -&gt;&gt; &apos;insightType&apos;::text)
   -&gt;  Bitmap Heap Scan on event_logs  (cost=27.33..387.08 rows=109 width=53) (actual time=0.068..0.203 rows=122 loops=1)
         Recheck Cond: (name = ANY (&apos;{InsightAddition,InsightEdit,InsightRemoval,InsightHover,InsightUICustomization,InsightDataPointClick}&apos;::text[]))
         Heap Blocks: exact=95
         -&gt;  Bitmap Index Scan on event_logs_name  (cost=0.00..27.30 rows=109 width=0) (actual time=0.049..0.050 rows=122 loops=1)
               Index Cond: (name = ANY (&apos;{InsightAddition,InsightEdit,InsightRemoval,InsightHover,InsightUICustomization,InsightDataPointClick}&apos;::text[]))
 Planning Time: 0.159 ms
 Execution Time: 0.315 ms
(9 rows)
</pre>
</details>

**Review by ebrodymoore — APPROVED** [review]
This is ready to be merged. The prod db has been updated!

**Pull request merged**